subsys/net/lib/lwm2m/lwm2m_observation.c

```diff
@@ -1508,38 +1510,163 @@ int lwm2m_engine_add_path_to_list(sys_slist_t *lwm2m_path_list, sys_slist_t *lwm
 				return 0;
 			}
 
-			if (entry->path.obj_id > path->obj_id) {
-				/* New entry have smaller Object ID */
-				add_before_current = true;
-			} else if (entry->path.obj_id == path->obj_id &&
-				   entry->path.level > path->level) {
-				add_before_current = true;
-			} else if (entry->path.obj_id == path->obj_id &&
-				   entry->path.level == path->level) {
-				if (path->level >= LWM2M_PATH_LEVEL_OBJECT_INST &&
-				    entry->path.obj_inst_id > path->obj_inst_id) {
-					/*
-					 * New have same Object ID
-					 * but smaller Object Instance ID
-					 */
-					add_before_current = true;
-				} else if (path->level >= LWM2M_PATH_LEVEL_RESOURCE &&
-					   entry->path.obj_inst_id == path->obj_inst_id &&
-					   entry->path.res_id > path->res_id) {
-					/*
-					 * Object ID and Object Instance id same
-					 * but Resource ID is smaller
-					 */
-					add_before_current = true;
-				} else if (path->level >= LWM2M_PATH_LEVEL_RESOURCE_INST &&
-					   entry->path.obj_inst_id == path->obj_inst_id &&
-					   entry->path.res_id == path->res_id &&
-					   entry->path.res_inst_id > path->res_inst_id) {
-					/*
-					 * Object ID, Object Instance id & Resource ID same
-					 * but Resource instance ID is smaller
-					 */
-					add_before_current = true;
+			if (new_entry->path.level < entry->path.level) {
```

Comment:
I wonder if we could simplify the decision making algorithm a bit, I don't think we need to have loops involved. I've tried the following and it seems to pass the new tests:
```c
switch (new_entry->path.level) {
case LWM2M_PATH_LEVEL_OBJECT:
	add_before_current =
		entry->path.obj_id >= new_entry->path.obj_id;
	break;

case LWM2M_PATH_LEVEL_OBJECT_INST:
	add_before_current =
		(entry->path.obj_id > new_entry->path.obj_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_OBJECT_INST &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id >= new_entry->path.obj_inst_id);
	break;

case LWM2M_PATH_LEVEL_RESOURCE:
	add_before_current =
		(entry->path.obj_id > new_entry->path.obj_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_OBJECT_INST &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id > new_entry->path.obj_inst_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_RESOURCE &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id == new_entry->path.obj_inst_id &&
		 entry->path.res_id >= new_entry->path.res_id);
	break;

case LWM2M_PATH_LEVEL_RESOURCE_INST:
	add_before_current =
		(entry->path.obj_id > new_entry->path.obj_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_OBJECT_INST &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id > new_entry->path.obj_inst_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_RESOURCE &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id == new_entry->path.obj_inst_id &&
		 entry->path.res_id > new_entry->path.res_id) ||

		(entry->path.level >= LWM2M_PATH_LEVEL_RESOURCE_INST &&
		 entry->path.obj_id == new_entry->path.obj_id &&
		 entry->path.obj_inst_id == new_entry->path.obj_inst_id &&
		 entry->path.res_id == new_entry->path.res_id &&
		 entry->path.res_inst_id >= new_entry->path.res_inst_id);
	break;
}
```
The assumption here is that the list is correctly ordered (which is a goal here) and there are no duplicates (which is checked earlier).

The general idea here is that we skip the initial path-level comparison and only focus on the new entry path-level. The rule for insertion at each level is that:
* If the ID of the existing entry is higher than the new one at lower path-level, insert,
* If the ID of the existing entry is higher or equal than the new one for the analyzed path-level, insert. The equal part takes care of entries with higher path-level, which may already be present in the list.

What do you think?

Reply:
I did wonder the same, the new adapted / fixed variant seems way too complex to me, but I didn't have time last year to further dig into it and simplify it. If it passes the tests, then I think we should be pretty good, I tried to construct test cases that should cover most of the difficult cases. 

I'll give your suggestion a more detailed look tomorrow and will come back to you

Reply:
I have looked into it also came to the conclusion that the algorithm should be sound given the assumptions. 

The last condition at each level, which is using `>=` instead of `>` compared to all the levels before is essential that special case for when all the levels match. 
I have updated the PR with the suggestion and made some minor changes (flipping the last comparison, at least for me it is easier to read & reason that way). 
And I documented the assumptions that we now rely on. Checking the precondition / invariant at each method invocation seemed to expensive.